lib/amqp/gen/producer.ex

```diff
@@ -126,26 +134,78 @@ defmodule Amqpx.Gen.Producer do
   end
 
   def handle_call(
-        {:publish, {exchange, routing_key, payload, options}},
+        {:publish, {exchange, routing_key, payload, options, retry_attempt}},
         _from,
         %{
           channel: channel,
           publisher_confirms: publisher_confirms,
-          publish_timeout: publish_timeout
+          publish_timeout: publish_timeout,
+          publish_retry_options: publish_retry_options
         } = state
       ) do
-    with :ok <-
-           Basic.publish(
-             channel,
-             exchange,
-             routing_key,
-             payload,
-             Keyword.merge([persistent: true], options)
-           ),
-         {:confirm, true} <-
-           {:confirm, confirm_delivery(publisher_confirms, publish_timeout, channel)} do
-      {:reply, :ok, state}
-    else
+    retry_policy = Keyword.get(publish_retry_options, :retry_policy, %{})
+    max_retries = Keyword.get(publish_retry_options, :max_retries, @default_max_retries)
+    backoff = Keyword.get(publish_retry_options, :backoff, @default_backoff)
+
+    publish_options = %{
+      publisher_confirms: publisher_confirms,
+      publish_timeout: publish_timeout,
+      max_retries: max_retries,
+      retry_policy: retry_policy,
+      backoff: backoff
+    }
+
+    result =
+      case retry_policy do
+        [] ->
+          do_publish(
+            channel,
+            exchange,
+            routing_key,
+            payload,
+            publish_options,
+            Keyword.merge([persistent: true], options)
+          )
+
+        _any ->
+          do_retry_publish(
+            channel,
+            exchange,
+            routing_key,
+            payload,
+            retry_attempt,
+            publish_options,
+            Keyword.merge([persistent: true], options)
+          )
+      end
+
+    handle_publish_result(result, state)
+  end
+
+  # Private functions
+  def do_publish_by(producer_name, exchange_name, routing_key, payload, options, attempt) do
+    case GenServer.call(producer_name, {:publish, {exchange_name, routing_key, payload, options, attempt}}) do
+      :ok ->
+        :ok
+
+      {:retry, next_attempt, backoff_time, _} ->
+        :timer.sleep(backoff_time)
```

Comment:
non mi piace tanto il timer sleep sincrono perché nel frattempo non puoi pubblicare nient'altro, magari si può restituire un `{:noreply, ...}` e usare una `:timer.send_after` per gestire il backoff

Reply:
@mbusi grazie dell'input! L'idea del tenere lo sleep nel **client** è stata pensata per evitare che la somma dei backoff che il _producer_ finirebbe ad aspettare prima di mandar la risposta non ecceda il timeout delle varie `GenServer.call` . In questo modo il producer rimane libero di pubblicare altro, e il client rimane sincrono (com'è attualmente) ma senza incorrere in timeout
Se abbiamo interpretato bene la tua soluzione, usare `{:noreply, ...}` nel producer e rispondere solo dopo le varie `send_after` rischierebbe di incorrere nel problema di timeout (una singola `.call` rimarrebbe appesa per tutto il tempo dei retry) o di costringere l'utente della publish_by a far calcoli astrusi per settare un timeout

In questo modo invece ci sembra di salvar capra e cavoli mantenendo lo stesso comportamento attuale.
Che ne pensi? (disponibili in chat/discord ovviamente per chiarimenti) :pray:

Reply:
Direi che anche con la situazione attuale puoi fare saltare il timeout della GenServer.call, semplicemente te ne parte solo uno (quella della singola chiamata appesa) invece che più di uno perché non puoi gestire chiamate in paralello così. Per altro visto che abbiamo anche il controllo del codice del client possiamo decidere lato amqpx cosa usare come timeout, quindi mi sembra un non problema sinceramente; non mi sembra sicuramente un trade-off valido per disabilitare completamente la gestione di possibili chiamate in parallelo quando si triggera un retry

Reply:
Visto che mi hai pingato su slack... siamo in uno stallo alla messicana? :D
Se vuoi parlarne live lanciami pure un calendar, se no penso che possiamo andare dritti con la tua implementazione e nel caso affrontare i problemi di concorrenza nel caso emergano... :shrug: